### ISSUE TITLE
qr doesn't work right.

### ISSUE THREAD

**Issue by philsupertramp** [philsupertramp/game-math#55]
in:
```
A
[
    2, 3
    3, 2
]
```

out:
```
Q
[
	-0.55470019622522893, 0.8320502943378435
	-0.83205029433784361, -0.55470019622522926
]

R
[
	-3.6055512754639887, -3.3282011773513744
	0, 1.3867504905630725
]
```

expected:
```
A = {{2, 3}, {3, 2}};
Q = { { -1, 1 }, {-1, -1} };
R = {{-4, -3}, {0, 1}};
```

**Comment by philsupertramp** [comment]
Let's try different implementation.
Example in python:
```python
import numpy as np

def qr(A):
    """Perform QR decomposition of matrix A using Gram-Schmidt process."""
    (num_rows, num_cols) = np.shape(A)

    Q = np.empty([num_rows, num_rows])
    col = 0

    # Compute orthogonal matrix Q.
    for a in A.T:
        u = np.copy(a)
        for i in range(0, col):
            proj = np.dot(np.dot(Q[:, i].T, a), Q[:, i])
            u -= proj

        e = u / np.linalg.norm(u)
        Q[:, col] = e
        col += 1

    return (Q, np.dot(Q.T, A))
```

**Comment by philsupertramp** [comment]
Translated into c++ code
```c++

/** Perform QR decomposition of matrix A using Gram-Schmidt process. 
 **/
std::tuple<Matrix<double>, Matrix<double>> qr(const Matrix<double>& A) {
  size_t num_rows = A.rows();
  size_t num_cols = A.columns();
  
  auto Q = zeros(num_rows, num_rows) ;
  size_t col = 0;
  
  for(size_t i_a = 0; i_a < num_cols; i_a++) {
    auto a = A.GetSlice(0, num_rows - 1, i_a, i_a);
    auto u = a;
    for(size_t i = 0; i < col; i++) {
      auto q = Q.GetSlice(0, num_rows, i, i);
      auto proj = q.Transpose() * a * q;
      u -= proj;
    } 
    auto e = u * 1.0/norm(u);
    Q.SetSlice(0, num_rows, col, col);
    col += 1;
  } 
  return std::make_pair(Q, Q.Transpose() * A);
}
```

**Comment by philsupertramp** [comment]
fixed in 0ce0597
